controllers/kustomization_controller.go

```diff
@@ -645,28 +650,41 @@ func (r *KustomizationReconciler) apply(ctx context.Context, kustomization kusto
 			}
 
 			cmd = fmt.Sprintf("%s --token %s", cmd, saToken)
+			diffcmd = fmt.Sprintf("%s --token %s", diffcmd, saToken)
+		}
+	}
+
+	// The reason for we parse the output via kubectl diff could be check from : https://github.com/fluxcd/kustomize-controller/issues/305
+	diffcommand := exec.CommandContext(applyCtx, "/bin/sh", "-c", diffcmd)
+	diffoutput, err := diffcommand.CombinedOutput()
+
+	// The reason that we still parse diff output when exit is for : https://github.com/kubernetes/kubernetes/pull/82336/files
+	// For the "diff" command Exit status:
+	//		 0
+	//		No differences were found.
+	//		 1
+	//		Differences were found.
+	//		 >1
+	//		Kubectl or diff failed with an error.
+	if differr, ok := err.(*exec.ExitError); ok {
+		if differr.ExitCode() > 1 {
+			log.Info("Kubectl or diff failed with an error, will execute apply soon")
+			output, applyErr := execApply(ctx, cmd)
+			if applyErr != nil {
+				return "", applyErr
+			}
+			resources = parseApplyOutput(output)
+		} else if differr.ExitCode() == 1 {
+			resources = parseDiffOutput(diffoutput)
+			//	Since we found difference in "diff" command, so we need to execute apply
+			log.Info("Differences found, will execute apply soon")
+			_, applyErr := execApply(ctx, cmd)
+			if applyErr != nil {
+				return "", applyErr
+			}
 		}
 	}
 
-	command := exec.CommandContext(applyCtx, "/bin/sh", "-c", cmd)
-	output, err := command.CombinedOutput()
-	if err != nil {
-		if errors.Is(applyCtx.Err(), context.DeadlineExceeded) {
-			return "", fmt.Errorf("apply timeout: %w", applyCtx.Err())
-		}
-
-		if string(output) == "" {
-			return "", fmt.Errorf("apply failed: %w, kubectl process was killed, probably due to OOM", err)
-		}
-
-		applyErr := parseApplyError(output)
-		if applyErr == "" {
-			applyErr = "no error output found, this may happen because of a timeout"
-		}
-		return "", fmt.Errorf("apply failed: %s", applyErr)
-	}
-
-	resources := parseApplyOutput(output)
 	log.Info(
```

Comment:
We shouldn't log that we applied something on the cluster if we didn't, please move the log below inside `else if differr.ExitCode() == 1 `

Reply:
Fix done